README.md
# Ubidots FONA Library 

**This library requires Arduino v1.6.0 or higher**

This is a library for the Adafruit FONA Cellular GSM Breakouts:

  * https://www.adafruit.com/products/1946
  * https://www.adafruit.com/products/1963
  * http://www.adafruit.com/products/2468
  * http://www.adafruit.com/products/2542

These modules use TTL Serial to communicate, so at least 2 pins are required to interface the FONA with your Arduino board.

## Required items

* Arduino UNO
* FONA
* An active SIM card **with a data plan**
* Ubidots_FONA library

## Pin Connections FONA -> Arduino

* Vio –> 5V (or 3V, if your Arduino has 3V logic)
* GND –> GND
* FONA_RX –> Pin 2
* FONA_TX –> Pin 3
* FONA_RST –> Pin 4
* FONA_KEY –> Pin 7
* FONA_PS –> Pin 8

## Run the examples!

1. Download the Ubidots_FONA library [here](https://github.com/ubidots/Ubidots-FONA/archive/1.2.0.zip)
2. Go to the Arduino IDE, click on Sketch -> Include Library -> Add .ZIP Library
3. Select the Ubidots_FONA .ZIP file and then "Accept" or "Choose"
4. Do the same with the Adafruit_FONA library.
5. Close the Arduino IDE and open it again.
**Before continuing, make sure your device can connect to the Internet! You can run Adafruit's example "FONAtest.ino" contained in [Adafruit's FONA](https://github.com/adafruit/Adafruit_FONA_Library/archive/1.3.0.zip) library to make sure your FONA is working properly and is able to connect to your mobile network.**
7. Now go to File -> Examples -> Ubidots FONA library and select one example.
6. Put your Ubidots token and variable ID
7. Put your mobile operator's APN settings (APN, USER, PASSWORD). You should be able to easily find your operator's APN settings on Google or in the operator's website.
8. Upload the code, open the Serial monitor to check the results. If no response is seen, try unplugging your Arduino and then plugging it again. Make sure the baud rate of the Serial monitor is set to the same one specified in your code.

Once you see some activity, go to your Ubidots account to check your data! 

To manually install the libraries, copy the Ubidots_FONA folder contained in the ZIP file into your *arduinosketchfolder*/libraries/ folder. You may need to create the libraries subfolder if its your first library, then restart the IDE.

If you're not sure how to add a library to the Arduino IDE please click [here](https://www.arduino.cc/en/Guide/Libraries) to see the steps as explained by Arduino.

You can contact support@ubidots.com for any inquiries, please specify your Mobile operator and provide the dump of the Serial Monitor for debugging purposes.

## Included Examples

### Send a value with latitude and longitude

To send a value with latitude and longitude use the example **UbidotsSendValuesWithContext** or just copy the following code:

```c

#include <Ubidots_FONA.h>
#include <SoftwareSerial.h>


#define APN  ""  // The APN of your operator
#define USER ""  // if your apn doesnt have username just leave it ""
#define PASS ""  // if your apn doesnt have password just leave it ""
#define TOKEN "Your_Token_Here"  // Replace it with your Ubidots token

Ubidots client(TOKEN);

void setup() {
  Serial.begin(115200);
  delay(2000);
  client.setApn(APN, USER, PASS);
}
void loop() {
    float value1 = analogRead(A0);
    float value2 = analogRead(A1);
    char context[25];
    char context_2[45];
    sprintf(context, "lat=%s$lng=%s", char(1.1232), char(4323.123));
    // To send latitude and longitude to Ubidots and see it in a map widget
    // The format of the context is like this, you must send it like this example
    sprintf(context_2, "Color=RED$Position=Down$Quantity=3");
    // You can send multiple context in one variable, to send it you must add a "$" symbol between every context
    client.add("Temperature", value1, context);  // Change for your variable name
    client.add("Humedity", value2,  context_2);
    client.sendAll();
    delay(5000);
}
```


## Get a value from a Ubidots variable

To get a value you could use the example inside the library or copy the following code:

```c
#include <Ubidots_FONA.h>
#include <SoftwareSerial.h>


#define APN  ""  // The APN of your operator
#define USER ""  // if your apn doesnt have username just leave it ""
#define PASS ""  // if your apn doesnt have password just leave it ""
#define TOKEN "Your_token_here"  // Replace it with your Ubidots token
#define ID "Your_id_here" // Replace it with your Ubidots' variable ID


Ubidots client(TOKEN);  
  
void setup() {
  Serial.begin(115200);
  delay(2000);
  client.setApn(APN, USER, PASS);
}

void loop() {
  float value = client.getValue(ID);  
  delay(1000); 
  Serial.println(value);
}
```


ChangeLog.md
# Change Log
All notable changes to this project will be documented in this file.

## [1.2.0] - 2016-05-26
### Changed
- We changed old examples of save value and get value to now examples using new functions of the library.
- Readme to explain how to send latitude and longitude in a value.
 
### Added
- New examples using shortert message to optimize the data plan.
- Possibility to send latitude and logitude in a variable.
- Feature to send multiple values in one request.
- Feature to send value using variable label.
- Feature to create a defaul data source with the code.

### Fixed
- Infinite loops.
- Multiple "If" in some functions.
- Reduced the size of message to send values.

### Removed
- Removed Adafruit FONA library of Required Items
- 

## [1.1.0] - 2016-01-28
### Added
- Changed old code to a new library with name Ubidots_FONA library


[1.2.0]: https://github.com/ubidots/Ubidots-FONA/tree/1.2.0
[1.1.0]: https://github.com/ubidots/Ubidots-FONA/tree/1.1.0

UbidotsFONA.h
/*
Copyright (c) 2013-2016 Ubidots.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Made by Mateo Velez - Metavix for Ubidots Inc

*/

#ifndef __UbidotsFONA_H_
#define __UbidotsFONA_H_
#define DEBUG_UBIDOTS

#include <SoftwareSerial.h>
#include <stdlib.h>
#include "Arduino.h"

#define DEFAULT_BUFFER_SIZE      64
#define FONA_RX 2
#define FONA_TX 3
#define FONA_RST 4
#define FONA_KEY 7
#define FONA_PS 8
#define SERVER "translate.ubidots.com"
#define PORT "9010"
#define USER_AGENT "FONA"
#define VERSION "/1.1"
#define MAX_VALUES 5


typedef struct Value {
    char  *idName;
    char  *contextOne;
    float idValue;
} Value;


class Ubidots{
private:
    SoftwareSerial fonaSS = SoftwareSerial(FONA_TX, FONA_RX);
    bool SendMessageAndwaitForOK(char* message, uint16_t timeout=4000);
    bool waitForMessage(const char *msg, uint32_t ts_max);
    bool isTimedOut(uint32_t ts) { return (long)(millis() - ts) >= 0; }
    int readLine(uint32_t ts_max);
    char* _token;
    char* _dsName;
    char* _dsTag;
    char* _server;
    char buffer[DEFAULT_BUFFER_SIZE];
    uint8_t currentValue;
    Value * val;

    
public:

    Ubidots(char* token, char* server = SERVER);
    void begin();
    void setDataSourceName(char *dsName);
    void setDataSourceTag(char *dsTag);
    void add(char *variable_id, float value, char *ctext1 = NULL);
    bool sendAll();
    float getValueWithDatasource(char* dsTag, char* idName);
    bool setApn(char* apn, char* user = "", char* pwd = "");
    void turnOffFona();
    void turnOnFona();
};
#endif



Ubidots_FONA.h
/*
Copyright (c) 2013-2016 Ubidots.

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

Made by Mateo Velez - Metavix for Ubidots Inc

*/

#ifndef __Ubidots_FONA_H_
#define __Ubidots_FONA_H_
#define DEBUG_UBIDOTS

#include <SoftwareSerial.h>
#include <stdlib.h>
#include "Arduino.h"

#define DEFAULT_BUFFER_SIZE      64
#define FONA_RX 2
#define FONA_TX 3
#define FONA_RST 4
#define FONA_KEY 7
#define FONA_PS 8
#define SERVER "translate.ubidots.com"
#define PORT "9010"
#define USER_AGENT "FONA"
#define VERSION "1.2"
#define MAX_VALUES 5


typedef struct Value {
    char  *idName;
    char  *contextOne;
    float idValue;
} Value;


class Ubidots{
private:
    SoftwareSerial fonaSS = SoftwareSerial(FONA_TX, FONA_RX);
    bool sendMessageAndwaitForOK(char* message, uint16_t timeout = 4000);
    bool waitForMessage(const char *msg, uint32_t ts_max);
    bool isTimedOut(uint32_t ts) { return (long)(millis() - ts) >= 0; }
    int readLine(uint32_t ts_max, bool multiline = false);
    bool begin();
    char* _token;
    char* _dsName;
    char* _dsTag;
    char* _server;
    char buffer[DEFAULT_BUFFER_SIZE];
    uint8_t currentValue;
    Value * val;
    // Variables to retro-compatibility
    char* _apn;
    char* _user;
    char* _pwd;
    bool httpTerm();
    bool httpInit();
    void flushInput();

public:
    Ubidots(char* token, char* server = SERVER);
    void setDataSourceName(char *dsName);
    void setDataSourceTag(char *dsTag);
    void add(char *variable_id, float value, char *ctext1 = NULL);
    bool sendAll();
    float getValueWithDatasource(char* dsTag, char* idName);
    bool setApn(char* apn, char* user = "", char* pwd = "");
    bool checkFona();
    // Deprecated functions
    void turnOffFona();
    void turnOnFona();
    void gprsNetwork(const __FlashStringHelper *apn, const __FlashStringHelper *username, const __FlashStringHelper *password);
    bool saveValue(char* myid, float value);
    float getValue(char* myid);
    void flushSerial();
    void gprsOnFona();
    
};
#endif  // __Ubidots_FONA_H_




Ubidots_fona.h
#ifndef __Ubidots_fona_H_
#define __Ubidots_fona_H_
#define DEBUG_UBIDOTS

#include <SoftwareSerial.h>
#include "Adafruit_FONA.h"
#include <stdlib.h>
#include "Arduino.h"

#define FONA_RX 2
#define FONA_TX 3
#define FONA_RST 4
#define FONA_KEY 7
#define FONA_PS 8

class Ubidots{
    private:
        SoftwareSerial fonaSS = SoftwareSerial(FONA_TX, FONA_RX);
        Adafruit_FONA fona = Adafruit_FONA(FONA_RST);
        char* _token;
        char* apn;

    public:
        
        Ubidots(char* token);
        void gprsNetwork(const __FlashStringHelper *apn, const __FlashStringHelper *username, const __FlashStringHelper *password);
        void save_value(float value, char* myid);
        float get_value(char* myid);
        void flushSerial();
        void gprsOnFona();
        void turnOffFona();
        void turnOnFona();
        void checkFona();
};
#endif



Ubidots_fona.cpp
#include "Ubidots_fona.h"
#include <stdlib.h>
/**
 * Constructor.
 */
 
Ubidots::Ubidots(char* token){
    _token = token;
    fonaSS.begin(4800);
}
void Ubidots::gprsNetwork(const __FlashStringHelper *apn, const __FlashStringHelper *username, const __FlashStringHelper *password){
        fona.setGPRSNetworkSettings(apn, username, password);
}
void Ubidots::save_value(float value, char* myid){
    uint16_t statuscode;
    int16_t length;
    char* url = (char *) malloc(sizeof(char) * 400);
    char* data = (char *) malloc(sizeof(char) * 50);
    char* val = (char *) malloc(sizeof(char) * 8);
    String(value).toCharArray(val,9);
    sprintf(url,"things.ubidots.com/api/v1.6/variables/%s/values?token=%s", myid, _token);
    sprintf(data,"{\"value\":%s}", val);  
    while (!fona.HTTP_POST_start(url, F("application/json"), (uint8_t *) data, strlen(data), &statuscode, (uint16_t *)&length)) {
        Serial.println("Failed!");
    }
    free(url);
    free(data);
    free(val);
    fona.HTTP_POST_end();
    flushSerial();
}
float Ubidots::get_value(char* myid){
    uint16_t statuscode;
    int16_t length;
    float num;
    char* url = (char *) malloc(sizeof(char) * 400);
    int i = 0;
    sprintf(url,"things.ubidots.com/api/v1.6/variables/%s/values?token=%s&page_size=1", myid, _token);
    while (!fona.HTTP_GET_start(url, &statuscode, (uint16_t *)&length)) {
        Serial.println("Failed!");
    }
    while(length>0){
        while(fonaSS.available()){
            url[i] = fonaSS.read();
            i++;
            length=length-1;
        }
        delay(10);
    }
    char* pch = strstr(url,"\"value\":");
    String raw(pch);
    int bodyPosinit =9+ raw.indexOf("\"value\":");
    int bodyPosend = raw.indexOf(", \"timestamp\"");
    raw.substring(bodyPosinit,bodyPosend).toCharArray(url, 10);
    num = atof(url);
    free(url);
    fona.HTTP_GET_end();
    flushSerial();
    return num;  
}
void Ubidots::flushSerial() {
    while (Serial.available())
    Serial.read();
}
void Ubidots::checkFona(){
    // See if the FONA is responding
    if (! fona.begin(fonaSS)) {           // can also try fona.begin(Serial1)
        Serial.println("Couldn't find FONA");
    }
    Serial.println("FONA is OK");
    //configure a GPRS APN
    
}
void Ubidots::gprsOnFona(){
    while(!fona.enableGPRS(true));
    Serial.println(F("Turn on"));
}
void Ubidots::turnOnFona(){
    Serial.println("Turning on Fona: ");
    while(digitalRead(FONA_PS)==LOW){
        digitalWrite(FONA_KEY, LOW);
    }
    digitalWrite(FONA_KEY, HIGH);
    delay(4000);
}

void Ubidots::turnOffFona(){
    Serial.println("Turning off Fona ");
    while(digitalRead(FONA_PS)==HIGH)
    {
        digitalWrite(FONA_KEY, LOW);
    }
    digitalWrite(FONA_KEY, HIGH);
    delay(4000);
}

Ubidots_FONA.cpp

#include <Stream.h>
#include <avr/pgmspace.h>
#ifdef ARDUINO_ARCH_AVR
#include <avr/wdt.h>
#else
#define wdt_reset()
#endif
#include <stdlib.h>

#include "Ubidots_FONA.h"
/**
 * Constructor.
 */
 
Ubidots::Ubidots(char* token, char* server){
    _token = token;
    _server = server;
    _dsName = NULL;
    _dsTag = "FONA";
    currentValue = 0;
    val = (Value *)malloc(MAX_VALUES*sizeof(Value));
}

bool Ubidots::begin() {
    pinMode(FONA_RST, OUTPUT);
    digitalWrite(FONA_RST, HIGH);
    delay(500);
    digitalWrite(FONA_RST, LOW);
    delay(500);
    digitalWrite(FONA_RST, HIGH);
    delay(500);
    Serial.println(F("Attempting to open comm with ATs"));
    int16_t timeout = 10000;
    while (timeout > 0) {
        while (fonaSS.available()) Serial.println(fonaSS.read());
        if (sendMessageAndwaitForOK("AT"))
            break;
        while (fonaSS.available()) Serial.println(fonaSS.read());
        fonaSS.println("AT");
        if (waitForMessage("AT", 5000)) 
            break;
        delay(500);
        timeout-=500;
    }
    if (timeout <= 0) {
        Serial.println(F("Timeout: No response to AT... last ditch attempt."));
        sendMessageAndwaitForOK("AT");
        delay(100);
        sendMessageAndwaitForOK("AT");
        delay(100);
        sendMessageAndwaitForOK("AT");
        delay(100);
    }
    sendMessageAndwaitForOK("ATE0");
    delay(100);
}

void Ubidots::setDataSourceName(char* dsName) {
    _dsName = dsName;
}

void Ubidots::setDataSourceTag(char* dsTag) {
    _dsTag = dsTag;
}

void Ubidots::add(char *variable_id, float value, char *ctext1) {
    (val+currentValue)->idName = variable_id;
    (val+currentValue)->contextOne = ctext1;
    (val+currentValue)->idValue = value;
    currentValue++;
    if (currentValue > MAX_VALUES) {
        currentValue = MAX_VALUES;
    }
}

bool Ubidots::sendMessageAndwaitForOK(char* message, uint16_t timeout) {
    int len;
    fonaSS.println(message);
    uint32_t ts_max = millis() + timeout;
    while ((len = readLine(ts_max)) >= 0) {
        if (len == 0) {
            // Skip empty lines
            continue;
        }
        if (strcmp_P(buffer, PSTR("OK")) == 0) {
            return true;
        } else if (strcmp_P(buffer, PSTR("ERROR")) == 0) {
            return false;
        }
        // Other input is skipped.
    }
    return false;
}

bool Ubidots::setApn(char* apn, char* user, char* pwd) {
    checkFona();
    int i = 0;
    char message[10][50];
    sprintf(message[0], "AT");
    sprintf(message[1], "AT+CSQ");
    sprintf(message[2], "AT+CIPSHUT");
    sprintf(message[3], "AT+CGATT?");
    sprintf(message[4], "AT+SAPBR=3,1,\"CONTYPE\",\"GPRS\"");
    sprintf(message[5], "AT+SAPBR=3,1,\"APN\",\"%s\"",apn);
    sprintf(message[6], "AT+SAPBR=3,1,\"USER\",\"%s\"",user);
    sprintf(message[7], "AT+SAPBR=3,1,\"PWD\",\"%s\"",pwd);
    sprintf(message[8], "AT+SAPBR=1,1");
    sprintf(message[9], "AT+SAPBR=2,1");
    for(i = 0; i < 10; i++) {
        if (!sendMessageAndwaitForOK(message[i], 6000)) {
            Serial.print("Error with ");
            Serial.println(message[i]);
            return false;
        }
    }
    return true;
}

bool Ubidots::sendAll() {
    int i;
    char message[8][50];
    String all;
    String str;
    all = USER_AGENT;
    all += "/";
    all += VERSION;
    all += "|POST|";
    all += _token;
    all += "|";
    all += _dsTag;
    if (_dsName != NULL) {
        all += ":";
        all += _dsName;
    }
    all += "=>";
    for (i = 0; i < currentValue; ) {
        str = String(((val + i)->idValue), 2);
        all += String((val + i)->idName);
        all += ":";
        all += str;
        if ((val + i)->contextOne != NULL) {
            all += "$";
            all += String((val + i)->contextOne);
        }
        all += ","; 
        i++;
    }
    all += "|end";
    sprintf(message[0], "AT+CIPMUX=0");
    sprintf(message[1], "AT+CIPSTART=\"TCP\",\"%s\",\"%s\"", _server, PORT);
    sprintf(message[2], "AT+CIPSEND");
    sprintf(message[4], "AT+CIPCLOSE");
    sprintf(message[5], ">");
    sprintf(message[6], "SEND OK");
    sprintf(message[7], "CLOSE OK");
    for (i = 0; i < 2; i++) {
        if (!sendMessageAndwaitForOK(message[i], 6000)) {
            Serial.print("Error with ");
            Serial.println(message[i]);
            currentValue = 0;
            return false;
        }
    }
    for (i = 2; i < 8; i++) {
        if (i != 3) {
            fonaSS.println(message[i]);
        } else {
            fonaSS.write(all.c_str());
            fonaSS.write(0x1A);
        }
        if (!waitForMessage(message[i+3], 6000)) {
            Serial.print("Error with ");
            Serial.println(message[i]);
            currentValue = 0;
            return false;
        }
    }
    currentValue = 0;
    return true;
}

float Ubidots::getValueWithDatasource(char* dsTag, char* idName) {
    float num;
    int i = 0;
    char allData[300];
    String response;
    uint8_t bodyPosinit;
    sprintf(allData, "%s/%s|LV|%s|%s:%s|end\n\x1A", USER_AGENT, VERSION, _token, dsTag, idName);
    fonaSS.println("AT+CIPSEND");
    if (!waitForMessage(">", 6000)) {
        Serial.println("Error at CIPSEND");
        return false;
    }
    fonaSS.write(allData);
    if (!waitForMessage("|OK", 6000)) {
        Serial.println("Error sending the message");
        return false;
    }
    response = String(buffer);
    bodyPosinit = 3 + response.indexOf("OK|");
    response = response.substring(bodyPosinit);
    num = response.toFloat();
    currentValue = 0;
    free(allData);
    return num;
}

bool Ubidots::waitForMessage(const char *msg, uint32_t ts_max) {
    int len;
    while ((len = readLine(ts_max)) >= 0) {
        if (len == 0) {
            // Skip empty lines
            continue;
        }
        if (strncmp(buffer, msg, strlen(msg)) == 0) {
            return true;
        }
    }
    return false;         // This indicates: timed out
}
int Ubidots::readLine(uint32_t ts_max, bool multiline) {
    int replyidx = 0;
    while (ts_max--) {
        if (replyidx >= 254) {
            break;
        }
        while(fonaSS.available()) {
            char c =  fonaSS.read();
            Serial.print(c);
            if (c == '\r') continue;
            if (c == 0xA) {
            if (replyidx == 0)   // the first 0x0A is ignored
                continue;
            }
            if (!multiline) {
                ts_max = 0;         // the second 0x0A is the end of the line
                break;
            }
            buffer[replyidx] = c;
            replyidx++;
        }
        if (ts_max == 0) {
            break;
        }
        delay(1);
    }
    buffer[replyidx] = 0;  // null term
    Serial.println(buffer);
    return replyidx;
}

void Ubidots::turnOnFona(){
    Serial.println("Turning on Fona: ");
    while(digitalRead(FONA_PS)==LOW){
        digitalWrite(FONA_KEY, LOW);
    }
    digitalWrite(FONA_KEY, HIGH);
    delay(4000);
}

void Ubidots::turnOffFona(){
    Serial.println("Turning off Fona ");
    while(digitalRead(FONA_PS)==HIGH)
    {
        digitalWrite(FONA_KEY, LOW);
    }
    digitalWrite(FONA_KEY, HIGH);
    delay(4000);
}


// ------------------------------------------------------------------
// -------------------------WARNING----------------------------------
// ------------------------------------------------------------------
/* Deprecated functions
* Next functions was deprecated at version 1.2 of Fona Library
*/

void Ubidots::gprsNetwork(const __FlashStringHelper *apn, const __FlashStringHelper *username, const __FlashStringHelper *password) {
    checkFona();
    _apn = reinterpret_cast<char*>(const_cast<__FlashStringHelper*>(apn));
    _user = reinterpret_cast<char*>(const_cast<__FlashStringHelper*>(username));
    _pwd = reinterpret_cast<char*>(const_cast<__FlashStringHelper*>(password));
}
bool Ubidots::saveValue(char* myid, float value) {
    char data[25];
    String val;
    val = String(value, 2);
    httpTerm();
    sprintf(data,"{\"value\":%s}", val.c_str());
    Serial.println(data);
    httpInit();
    Serial.println(_token);
    fonaSS.print(F("AT+HTTPPARA=\"URL\",\"things.ubidots.com/api/v1.6/variables/"));
    fonaSS.print(myid);
    fonaSS.print(F("/values?token="));
    fonaSS.print(_token);
    if (!sendMessageAndwaitForOK("\"", 4000)) {
        Serial.println(F("Error with AT+HTTPARA URL"));     
        httpTerm();
        return false;
    }
    if (!sendMessageAndwaitForOK("AT+HTTPPARA=\"CONTENT\",\"application/json\"", 4000)) {
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Error with AT+HTTPARA CONTENT"));
#endif
        httpTerm();
        return false;
    }
    fonaSS.print(F("AT+HTTPDATA="));
    fonaSS.print(strlen(data));
    fonaSS.print(F(","));
    fonaSS.println(120000);
    if (!waitForMessage("DOWNLOAD",2000)) {
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Error with AT+HTTPDATA"));
#endif
        httpTerm();
        return false;
    }
    fonaSS.write(data, strlen(data));
    if (!waitForMessage("OK",2000)) {
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Error sending data and data length"));
#endif
        httpTerm();
        return false;
    }
    fonaSS.println(F("AT+HTTPACTION=1"));  // HTTPACTION=1 is a POST method
    delay(5000);
    if (!waitForMessage("+HTTPACTION:1,201", 2000)) {
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Error with AT+HTTPACTION=1"));
        Serial.println(F("Status = 603 is DNS Error, maybe your SIM doesn't have mobile data"));
        Serial.println(F("Status = 601 Network Error, maybe your SIM doesn't have mobile data"));
        Serial.println(F("Status = 400 Bad Request, Your URL is wrong"));
        Serial.println(F("Status = 402 Payment Required, You exceed your dots in Ubidots"));
#endif
        httpTerm();
        return false;
    } else {
        fonaSS.println(F("AT+HTTPREAD"));
        if (!waitForMessage("+HTTPREAD:", 2000)) {
#ifdef DEBUG_UBIDOTS
            Serial.println(F("Error with AT+HTTPREAD. Closing HTTP Client"));
#endif
            httpTerm();
            return false;
        }
    }
    httpTerm();
    return true;
}
float Ubidots::getValue(char* myid) {
    float num;
    String raw;
    uint8_t bodyPosinit;
    uint8_t bodyPosend;
    flushInput();
    httpTerm();
    httpInit();
    fonaSS.print(F("AT+HTTPPARA=\"URL\",\"things.ubidots.com/api/v1.6/variables/"));
    fonaSS.print(myid);
    fonaSS.print(F("/values?token="));
    fonaSS.print(_token);
    if (!sendMessageAndwaitForOK("&page_size=1\"", 6000)) {
        httpTerm();
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Error with AT+HTTPPARA URL. Closing HTTP Client"));
#endif      
        return 1;
    }
    fonaSS.println(F("AT+HTTPACTION=0"));
    if (!waitForMessage("+HTTPACTION:0,200", 4000)) {
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Status = 603 is DNS Error, maybe your SIM doesn't have mobile data"));
        Serial.println(F("Status = 601 Network Error, maybe your SIM doesn't have mobile data"));
        Serial.println(F("Status = 400 Bad Request, Your URL is wrong"));
        Serial.println(F("Status = 402 Payment Required, You exceed your dots in Ubidots"));
#endif
        httpTerm();
        return 1; 
    } else {
        fonaSS.println(F("AT+HTTPREAD"));
        if (!waitForMessage("+HTTPREAD:", 3000)) {
#ifdef DEBUG_UBIDOTS
        Serial.println(F("Error with AT+HTTPREAD. Closing HTTP Client"));
#endif
            httpTerm();
            return 1;
        }
        char* pch = strstr(buffer,"\"value\":");
#ifdef DEBUG_UBIDOTS
        String raw(pch);
#endif
        int bodyPosinit =9 + raw.indexOf("\"value\":");
        int bodyPosend = raw.indexOf(", \"timestamp\"");
        raw.substring(bodyPosinit, bodyPosend).toCharArray(buffer, 10);
        num = atof(buffer);
        httpTerm();
        return num;
    }
}
void Ubidots::gprsOnFona() {
    checkFona();
    setApn(_apn, _user, _pwd);
}
void Ubidots::flushSerial() {
    while (Serial.available())
    Serial.read();
}
bool Ubidots::httpTerm() {
    if (!sendMessageAndwaitForOK("AT+HTTPTERM", 6000)) {
        Serial.print("Error with AT+HTTPTERM");
        return false;
    }
    return true;
}
bool Ubidots::httpInit(){
    uint8_t i = 0;
    char message[3][50];
    sprintf(message[0], "AT+HTTPINIT");
    sprintf(message[1], "AT+HTTPPARA=\"CID\",1");
    sprintf(message[2], "AT+HTTPPARA=\"UA\",\"%s /%s\"");
    for (i = 0; i < 3; i++) {
        if (!sendMessageAndwaitForOK(message[i], 6000)) {
            Serial.print("Error with ");
            Serial.println(message[i]);
            return false;
        }
    }
    return true;
}

void Ubidots::flushInput() {
    // Read all available serial input to flush pending data.
    uint16_t timeoutloop = 0;
    while (timeoutloop++ < 40) {
        while(Serial.available()) {
            Serial.read();
            timeoutloop = 0;  // If char was received reset the timer
        }
        delay(1);
    }
}

bool Ubidots::checkFona() {
    fonaSS.begin(4800);
    void begin();
    if (!sendMessageAndwaitForOK("ATE0", 6000)) {
        Serial.print("Couldn't find FONA");
        return false;
        while (1);
    }
    return true;
}